assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js

```diff
@@ -32,83 +37,123 @@ import { CORE_MODULES } from '../../../googlesitekit/modules/datastore/constants
 import { KEY_METRICS_WIDGETS } from '../key-metrics-widgets';
 import MetricItem from './MetricItem';
 import useViewOnly from '../../../hooks/useViewOnly';
+import { Fragment } from '@wordpress/element';
 const { useSelect } = Data;
 
 export default function Metrics( { savedMetrics } ) {
 	const isViewOnlyDashboard = useViewOnly();
 
-	const availableMetrics = useSelect( ( select ) => {
-		const widgets =
-			select( CORE_WIDGETS ).getWidgets(
-				AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY
-			) || [];
-
-		const metrics = widgets
-			.filter( ( { slug } ) => savedMetrics.includes( slug ) )
-			.map( ( { slug } ) => slug );
-
-		const { isKeyMetricAvailable } = select( CORE_USER );
+	const { isKeyMetricAvailable } = useSelect( ( select ) =>
+		select( CORE_USER )
+	);
 
-		const { getModule } = select( CORE_MODULES );
+	const { getModule } = useSelect( ( select ) => select( CORE_MODULES ) );
 
-		return Object.keys( KEY_METRICS_WIDGETS )
-			.sort(
-				( a, b ) =>
-					metrics.includes( b ) - metrics.includes( a ) ||
-					metrics.indexOf( a ) - metrics.indexOf( b )
+	const displayInList = useSelect(
+		( select ) => ( metric ) =>
+			KEY_METRICS_WIDGETS[ metric ].displayInList(
+				select,
+				isViewOnlyDashboard
 			)
-			.reduce( ( acc, metric ) => {
-				if ( ! isKeyMetricAvailable( metric ) ) {
-					return acc;
-				}
+	);
+
+	const getWidget = useSelect(
+		( select ) => ( metric ) => select( CORE_WIDGETS ).getWidget( metric )
+	);
 
-				if (
-					typeof KEY_METRICS_WIDGETS[ metric ].displayInList ===
-						'function' &&
-					! KEY_METRICS_WIDGETS[ metric ].displayInList(
-						select,
-						isViewOnlyDashboard
-					)
-				) {
-					return acc;
+	const metricsListReducer = ( acc, metric ) => {
+		if ( ! isKeyMetricAvailable( metric ) ) {
+			return acc;
+		}
+
+		if (
+			typeof KEY_METRICS_WIDGETS[ metric ].displayInList === 'function' &&
+			! displayInList( metric )
+		) {
+			return acc;
+		}
+
+		const widget = getWidget( metric );
+
+		KEY_METRICS_WIDGETS[ metric ].disconnectedModules =
```

Comment:
I realise this code was simply moved in this PR, but this sort of side effect in a reducer is a bit of a warning flag - ideally a reducer function should be a pure function.

I don't think we should be modifying `KEY_METRICS_WIDGETS` directly at this point, rather we should be making a copy of `KEY_METRICS_WIDGETS[ metric ]` in our accumulator and adding `disconnectedModules` to this copy...